pkg/daemon/daemon.go

```diff
@@ -245,145 +348,80 @@ func (dn *Daemon) operatorConfigChangeHandler(old, new interface{}) {
 	// Get the latest NodeState
 	var latestState *sriovnetworkv1.SriovNetworkNodeState
 	err = wait.PollImmediate(10*time.Second, 1*time.Minute, func() (bool, error) {
 		latestState, lastErr = dn.client.SriovnetworkV1().SriovNetworkNodeStates(namespace).Get(context.Background(), dn.name, metav1.GetOptions{})
 		if lastErr == nil {
 			return true, nil
 		}
-		glog.Warningf("nodeStateChangeHandler(): Failed to fetch node state %s (%v); retrying...", dn.name, lastErr)
+		glog.Warningf("nodeStateSyncHandler(): Failed to fetch node state %s (%v); retrying...", dn.name, lastErr)
 		return false, nil
 	})
 	if err != nil {
 		if err == wait.ErrWaitTimeout {
-			dn.exitCh <- goerr.Wrapf(lastErr, "nodeStateChangeHandler(): Timed out trying to fetch the latest node state %s", dn.name)
-			return
+			return fmt.Errorf("nodeStateSyncHandler(): Timed out trying to fetch the latest node state: %v", lastErr)
 		}
-		dn.exitCh <- err
-		return
+		return fmt.Errorf("%v: %v", err, lastErr)
 	}
-	latest := latestState.GetObjectMeta().GetGeneration()
-	if latest != newState.GetObjectMeta().GetGeneration() {
-		glog.Errorf("nodeStateChangeHandler(): the lastet generation is %d, skip", latest)
-		return
+	latest := latestState.GetGeneration()
+	if latest != generation {
+		glog.Infof("nodeStateSyncHandler(): the latest generation is %d, skip", latest)
+		return nil
 	}
-	if latest == oldState.GetObjectMeta().GetGeneration() {
-		glog.V(2).Infof("nodeStateChangeHandler(): Interface not changed")
-		return
+
+	if dn.nodeState.GetGeneration() == latest {
+		glog.V(0).Infof("nodeStateSyncHandler(): Interface not changed")
+		return nil
 	}
+
 	dn.refreshCh <- Message{
 		syncStatus:    "InProgress",
 		lastSyncError: "",
 	}
+
+	// load plugins if has not loaded
+	if len(dn.LoadedPlugins) == 0 {
+		err = dn.loadVendorPlugins(latestState)
+		if err != nil {
+			glog.Errorf("nodeStateSyncHandler(): failed to load vendor plugin: %v", err)
+			return err
+		}
+	}
+
 	reqReboot := false
 	reqDrain := false
 	for k, p := range dn.LoadedPlugins {
-		d, r, err := p.OnNodeStateChange(oldState, latestState)
+		d, r := false, false
+		if dn.nodeState.GetName() == "" {
+			d, r, err = p.OnNodeStateAdd(latestState)
+		} else {
+			d, r, err = p.OnNodeStateChange(dn.nodeState, latestState)
+		}
 		if err != nil {
-			glog.Errorf("nodeStateChangeHandler(): plugin %s error: %v", k, err)
-			dn.exitCh <- err
-			return
+			glog.Errorf("nodeStateSyncHandler(): plugin %s error: %v", k, err)
+			return err
 		}
 		reqDrain = reqDrain || d
 		reqReboot = reqReboot || r
 	}
-	glog.V(2).Infof("nodeStateChangeHandler(): reqDrain %v, reqReboot %v", reqDrain, reqReboot)
+	glog.V(0).Infof("nodeStateSyncHandler(): reqDrain %v, reqReboot %v", reqDrain, reqReboot)
 
 	if reqDrain {
-		glog.Info("nodeStateChangeHandler(): drain node")
-		dn.drainNode(latestState.GetName())
+		glog.Info("nodeStateSyncHandler(): drain node")
+		if err := dn.drainNode(dn.name); err != nil {
+			return err
```

Comment:
When dn.drainNode fails, shall it call completeDrain and other nodes may get chance to start draining?
This node, when drainNode fails, will be marked with syncError (draining in this node completes with error).
This may avoid draining process be blocked by one error node.

Reply:
When draining fails, it most likely caused by a lack of spare resources. I don't think we shall continue to drain other nodes in such a case.

Reply:
Lack of resources in one node doesn't mean other nodes are not drainable. Take a fat pod as an example, it may not be evicted to another node due to resource constraints, but thin pods on other nodes might be able to migrate to this node or any other nodes(except the fat pod node).

Reply:
Before adding this patch, I think we agree on the need of a drain timeout, which allows the drain process on the node to proceed. 
After introducing this patch, if we don't complete the drain process upon failure in one node, it will stuck forever.

Reply:
There is a timeout in draining. If the draining fails, the event will be put back to the queue and retry. Yes, the node may be stuck in 'unschedulable' forever. But if we `uncordon` this node and give other nodes the opportunity to be drained, and the draining fails on other nodes. Then multiple nodes may be stuck in a drain-failed-uncordon cycle, which could affect more than one node.

Reply:
Just a note: the draining may or may not fail on other nodes, the failure of drain in the affected node does not necessarily affect non-sriov components from running on the same node or other nodes. 

It's that for current implementation, we choose to block other nodes from draining in case any one of cluster nodes fails.

Reply:
It would make sense to understand if the drain failed because pods could not be evicted (in that case it does not make sense to retry) or because of some error that is worth retrying (such as connection error with the apiserver).
But this would probably be too complex and error prone. 
I see @zshi-redhat 's point, if there's a fat pod, having some nodes drained or not is a matter of the order we receive the states. How about the initial idea of `completeDrain` and returning an error? In that case, on the long run if a node is not able to drain won't affect the others.
Also, note that in `drainNode` we don't check if the wait for the node to be drainable or not succeded or failed. Not sure it's wanted or not.